File: vaulto/src/utils/metadataExtractor.js
// ========================================
// 🚀 ADVANCED METADATA EXTRACTION ENGINE
// ========================================

// ----------------------------------------
// 1. Extract Store / Brand Name
// ----------------------------------------
export function extractStore(text) {
  const knownStores = [
    "Amazon", "Flipkart", "Myntra", "Ajio", "Zara", "H&M", "Starbucks",
    "Swiggy", "Zomato", "Uber", "Nike", "Adidas", "Decathlon", "Blinkit",
    "BigBasket", "Dominos", "Pizza Hut", "Nykaa"
  ];

  const cleanText = text.toLowerCase();

  for (const store of knownStores) {
    if (cleanText.includes(store.toLowerCase())) {
      return store;
    }
  }

  // fallback — first capitalized word (usually store name)
  const match = text.match(/\b[A-Z][A-Za-z]{3,}\b/);
  return match ? match[0] : "Unknown Store";
}

// ----------------------------------------
// 2. Extract Coupon Code
// ----------------------------------------
export function extractCouponCode(text) {
  // Common code patterns:
  // SAVE20, WELCOME100, GET50OFF, AB12CD, 20OFF, TRYNEW, etc.
  const patterns = [
    /\b[A-Z0-9]{5,12}\b/g,                   // ALL CAPS CODE ABC12345
    /\bSAVE[0-9]{1,3}\b/g,                   // SAVE20
    /\bGET[0-9]{1,3}\b/g,                    // GET50
    /\b[A-Z]{2,6}[0-9]{2,4}\b/g,             // WELCOME50
    /\b[0-9]{2}OFF\b/g                       // 20OFF
  ];

  for (const regex of patterns) {
    const match = text.match(regex);
    if (match) return match[0];
  }

  return "Not Detected";
}

// ----------------------------------------
// 3. Extract Discount
// ----------------------------------------
export function extractDiscount(text) {
  // Detect patterns: 20% OFF, FLAT ₹200 OFF, Buy 1 Get 1, etc.
  const percent = text.match(/\b([0-9]{1,3})%\s*off/i);
  if (percent) return percent[0];

  const flat = text.match(/₹\s*[0-9]{2,5}\s*off/i);
  if (flat) return flat[0];

  if (/buy\s*1\s*get\s*1/i.test(text)) return "Buy 1 Get 1 Free";
  if (/bogo/i.test(text)) return "BOGO Offer";

  return "Not Detected";
}

// ----------------------------------------
// 4. Extract Expiry Date
// ----------------------------------------
export function extractExpiry(text) {
  // Supports: 12/08/2025, 12-08-25, Aug 12 2025, Valid till 10th May, etc.
  const datePatterns = [
    /\b([0-9]{1,2}[\/\-][0-9]{1,2}[\/\-][0-9]{2,4})\b/,            // 12/08/2025
    /\b([A-Za-z]{3,9}\s+[0-9]{1,2}[,]?\s*[0-9]{2,4})\b/,            // Aug 12 2025
    /\b(valid till|valid until|expires on)\s*([A-Za-z0-9 ,\-]+)/i   // Valid till May 2025
  ];

  for (const regex of datePatterns) {
    const match = text.match(regex);
    if (match) return match[0];
  }

  return "No expiry found";
}

// ----------------------------------------
// 5. Detect Category
// ----------------------------------------
export function detectCategory(store, text) {
  const lower = (store + " " + text).toLowerCase();

  if (lower.match(/food|restaurant|pizza|burger|swiggy|zomato|dominos|starbucks/))
    return "Food";

  if (lower.match(/shirt|clothes|fashion|myntra|ajio|zara|h&m/))
    return "Fashion";

  if (lower.match(/electronics|mobile|laptop|amazon|flipkart/))
    return "Electronics";

  if (lower.match(/travel|flight|hotel|uber/))
    return "Travel";

  return "Others";
}

// ----------------------------------------
// 6. MAIN FUNCTION — EXTRACT EVERYTHING
// ----------------------------------------
export function extractMetadata(text) {
  const store = extractStore(text);
  const code = extractCouponCode(text);
  const expiry = extractExpiry(text);
  const discount = extractDiscount(text);
  const category = detectCategory(store, text);

  return {
    store,
    code,
    expiry,
    discount,
    category,
    rawText: text
  };
}

File: backend/src/main/java/com/vaulto/util/MetadataExtractor.java
package com.vaulto.util;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class MetadataExtractor {

    private static final String[] KNOWN_STORES = {
        "Amazon", "Flipkart", "Myntra", "Ajio", "Swiggy", "Zomato",
        "Uber", "Ola", "MakeMyTrip", "Goibibo", "Nykaa", "BigBasket",
        "Blinkit", "Zepto", "Dominos", "PizzaHut", "KFC", "McDonalds"
    };

    public static String extractStore(String text) {
        if (text == null) return "Unknown";
        for (String store : KNOWN_STORES) {
            if (Pattern.compile("(?i)\\b" + store + "\\b").matcher(text).find()) {
                return store;
            }
        }
        
        Matcher matcher = Pattern.compile("\\b([A-Z][a-z]+)\\b").matcher(text);
        if (matcher.find()) {
            return matcher.group(1);
        }
        return "Unknown";
    }

    public static String extractCouponCode(String text) {
        if (text == null) return "";
        Matcher matcher = Pattern.compile("\\b([A-Z0-9]{5,15})\\b").matcher(text);
        if (matcher.find()) {
            return matcher.group(1);
        }
        return "";
    }

    public static String extractDiscount(String text) {
        if (text == null) return "";
        
        Matcher percentMatcher = Pattern.compile("(\\d{1,2}%\\s*(?:OFF|off|Off))").matcher(text);
        if (percentMatcher.find()) return percentMatcher.group(1);

        Matcher flatMatcher = Pattern.compile("(?:₹|Rs\\.?)\\s*(\\d+)\\s*(?:OFF|off|Off)").matcher(text);
        if (flatMatcher.find()) return "₹" + flatMatcher.group(1) + " OFF";

        Matcher bogoMatcher = Pattern.compile("(?i)(buy\\s*1\\s*get\\s*1|bogo)").matcher(text);
        if (bogoMatcher.find()) return "BOGO";

        return "";
    }

    public static String extractExpiry(String text) {
        if (text == null) return "";
        
        Matcher dateMatcher = Pattern.compile("(\\d{2}[-/]\\d{2}[-/]\\d{2,4})").matcher(text);
        if (dateMatcher.find()) return dateMatcher.group(1);

        Matcher textDateMatcher = Pattern.compile("(?i)(?:valid till|expires on|expiry)\\s*:\\s*([A-Za-z]+\\s*\\d{1,2},?\\s*\\d{4})").matcher(text);
        if (textDateMatcher.find()) return textDateMatcher.group(1);

        return "";
    }

    public static String detectCategory(String store, String text) {
        String lowerStore = store.toLowerCase();
        String lowerText = text.toLowerCase();

        if (lowerStore.matches(".*(swiggy|zomato|dominos|pizzahut|kfc|mcdonalds|bigbasket|blinkit|zepto).*") || 
            lowerText.matches(".*(food|meal|grocery|restaurant|pizza|burger).*")) {
            return "Food & Groceries";
        }

        if (lowerStore.matches(".*(myntra|ajio|nykaa).*") || 
            lowerText.matches(".*(fashion|clothing|shoes|beauty|makeup).*")) {
            return "Fashion & Beauty";
        }

        if (lowerStore.matches(".*(amazon|flipkart).*") || 
            lowerText.matches(".*(electronics|mobile|laptop|gadget).*")) {
            return "Electronics";
        }

        if (lowerStore.matches(".*(makemytrip|goibibo|uber|ola).*") || 
            lowerText.matches(".*(travel|flight|hotel|cab|ride).*")) {
            return "Travel & Rides";
        }

        return "Others";
    }
}
